### PR TITLE
feat: add github graphql implementation

### PR DESCRIPTION
**Pull request by pritamstyz4ever**
## Context

We need to interact with github GraphQL API to get info on enterprise users

## Objective

This PR adds functions to interact with Github Graph QL API

## References
https://docs.github.com/en/enterprise-cloud@latest/graphql/reference/queries

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -0,0 +1,24 @@
+Copyright 2016 Yahoo Inc.
```
Comment (tkyi) [inline LICENSE]:
```suggestion
Copyright 2024 Yahoo Inc.
```

File: index.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        this.sdGql = new SDGraphQLClient({
+            graphqlUrl: config.graphQlUrl || 'https://api.github.com/graphql'
+        });
+    }
+
+    /**
+     * Gets the enterprise user account schema
+     * @param {String} config              The config object
+     * @param {String} config.slug         The github enterprise slug
+     * @param {String} config.username     The github username
```
Comment (tkyi) [inline index.js]:
```suggestion
     * @param {String} config.username     The github username
     * @param {String} config.token             Access token
```

File: sd-graphql/index.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        return res.body;
+    }
+
+    /**
+     * Run a graphql mutation and return the result
+     * @param {String} schema             The schema object
+     * @param {String} schema.mutation    The graphql mutation
+     * @param {Object} schema.variables   The variables to use in the mutation
+     * @param {Object} schema.token       The token to use for authentication
+     * @param {Object} options     The options to pass to the request
```
Comment (tkyi) [inline sd-graphql/index.js]:
```suggestion
     * @param {Object} schema.options     The options to pass to the request
```

File: index.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+     * @returns Array https://docs.github.com/en/enterprise-cloud@latest/graphql/reference/unions#enterprisemember
+     */
+    async listEnterpriseMembers(config) {
+        return this._listEnterpriseMembersHelper(config, null, []);
+    }
+
+    /**
+     * Gets the the github user schema
+     * @param {String} config    The config object
+     * @param {String} login     The github username
```
Comment (tkyi) [inline index.js]:
```suggestion
     * @param {Object} config             Config object
     * @param {String} config.login     Github username
     * @param {String} config.toke      Access token
```

File: test/index.test.js [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+const queries = require('../queries');
+
+sinon.assert.expose(assert, {
+    prefix: ''
+});
+
+describe('GithubGraphQL', () => {
+    let githubGql;
+    let GithubScmGraphQL;
+    // let sdGqlMock;
```
Comment (tkyi) [inline test/index.test.js]:
```suggestion
```

File: README.md [hunk 6]
```diff
@@ -1,2 +1,83 @@
...
+This method retrieves a list of enterprise members based on schema https://docs.github.com/en/enterprise-cloud@latest/graphql/reference/unions#enterprisemember.
+
+##### getUser
+Required parameters:
+
+| Parameter        | Type  |  Description |
+| :-------------   | :---- | :-------------|
+| schema     | Object | The schema object |
+| schema.login     | String | The github user's login name |
+| schema.token | String | some_token | The github token to interact with the graphql api |
```
Comment (tkyi) [inline README.md]:
```suggestion
| schema.token | String | The github token to interact with the graphql api |
```

File: README.md [hunk 7]
```diff
@@ -1,2 +1,83 @@
...
+Gets the enterprise user account based on schema https://docs.github.com/en/enterprise-cloud@latest/graphql/reference/objects#enterpriseuseraccount
+
+#### listEnterpriseMembers
+Required parameters:
+
+| Parameter        | Type  |  Description |
+| :-------------   | :---- | :-------------|
+| schema     | Object | The schema object |
+| schema.slug     | String | The github enterprise slug |
+| schema.token | String | some_token | The github token to interact with the graphql api |
```
Comment (tkyi) [inline README.md]:
```suggestion
| schema.token | String | The github token to interact with the graphql api |
```

File: README.md [hunk 8]
```diff
@@ -1,2 +1,83 @@
...
+
+#### getEnterpriseUserAccount
+Required parameters:
+
+| Parameter        | Type  |  Description |
+| :-------------   | :---- | :-------------|
+| schema     | Object | The schema object |
+| schema.slug     | String | The github enterprise slug |
+| schema.username      | String | The github user's login name |
+| schema.token | String | some_token | The github token to interact with the graphql api |
```
Comment (tkyi) [inline README.md]:
```suggestion
| schema.token | String | The github token to interact with the graphql api |
```

File: README.md [hunk 9]
```diff
@@ -1,2 +1,83 @@
...
+### Methods
+
+#### getEnterpriseUserAccount
+Required parameters:
+
+| Parameter        | Type  |  Description |
+| :-------------   | :---- | :-------------|
+| schema     | Object | The schema object |
+| schema.slug     | String | The github enterprise slug |
+| schema.username      | String | The github user's login name |
```
Comment (tkyi) [inline README.md]: is there a particular reason why this one is `username` vs `login`?

Comment (pritamstyz4ever) [reply README.md]: `login` is used in GQL context, I guess I can change this to `login`

File: sd-graphql/index.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    constructor(config) {
+        this._config = config;
+    }
+
+    /**
+     * Run a graphql query and return the result
+     * @param {String} schema              The schema object
+     * @param {String} schema.query        The graphql query
+     * @param {Object} schema.variables    The variables to use in the query
+     * @param {Object} schema.token        The token to use for authentication
```
Comment (tkyi) [inline sd-graphql/index.js]:
```suggestion
     * @param {String} schema.token        The token to use for authentication
```

File: sd-graphql/index.js [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+        return res.body;
+    }
+
+    /**
+     * Run a graphql mutation and return the result
+     * @param {String} schema             The schema object
+     * @param {String} schema.mutation    The graphql mutation
+     * @param {Object} schema.variables   The variables to use in the mutation
+     * @param {Object} schema.token       The token to use for authentication
```
Comment (tkyi) [inline sd-graphql/index.js]:
```suggestion
     * @param {String} schema.token       The token to use for authentication
```

File: test/sd-graphql.test.js [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+                user(login: $username) {
+                    name
+                    email
+                    login
+                }
+            }
+        `;
+
+        const variables = {
+            username: 'ai_humanoid'
```
Comment (tkyi) [inline test/sd-graphql.test.js]:
```suggestion
            query GetUser($login: String!) {
                user(login: $login) {
                    name
                    email
                    login
                }
            }
        `;

        const variables = {
            login: 'ai_humanoid'
```

**Review by tkyi — APPROVED** [review]

**Pull request merged**